instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameProvider.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.resources;
+
+import static java.util.logging.Level.FINE;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.ConditionalResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/** A {@link ResourceProvider} that will attempt to guess the application name from the jar name. */
+@AutoService(ResourceProvider.class)
+public final class JarServiceNameProvider implements ConditionalResourceProvider {
+
+  private static final Logger logger = Logger.getLogger(JarServiceNameProvider.class.getName());
+
+  private final Supplier<String[]> getProcessHandleArguments;
+  private final Function<String, String> getSystemProperty;
+  private final Predicate<Path> fileExists;
+
+  @SuppressWarnings("unused") // SPI
+  public JarServiceNameProvider() {
+    this(ProcessArguments::getProcessArguments, System::getProperty, Files::isRegularFile);
+  }
+
+  // visible for tests
+  JarServiceNameProvider(
+      Supplier<String[]> getProcessHandleArguments,
+      Function<String, String> getSystemProperty,
+      Predicate<Path> fileExists) {
+    this.getProcessHandleArguments = getProcessHandleArguments;
+    this.getSystemProperty = getSystemProperty;
+    this.fileExists = fileExists;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+    Path jarPath = getJarPathFromProcessHandle();
+    if (jarPath == null) {
+      jarPath = getJarPathFromSunCommandLine();
+    }
+    if (jarPath == null) {
+      return Resource.empty();
+    }
+    String serviceName = getServiceName(jarPath);
+    logger.log(FINE, "Auto-detected service name from the jar file name: {0}", serviceName);
+    return Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, serviceName));
+  }
+
+  @Override
+  public boolean shouldApply(ConfigProperties config, Resource existing) {
+    String serviceName = config.getString("otel.service.name");
+    Map<String, String> resourceAttributes = config.getMap("otel.resource.attributes");
```

Comment:
are these needed b/c this resource provider can't run after the built-in resource provider?

Reply:
do you think this is changeable, to make `otel.service.name` and `otel.resource.attributes` be real resource detectors, so that they can be leveraged as part of `shouldApply` (cc @jack-berg)

(I think this is a perfectly good workaround tho, and hopefully it's not a common use case)